doc/source/f2py/string_session.dat

```diff
@@ -1,27 +1,30 @@
 >>> import mystring
->>> print mystring.foo.__doc__
-foo - Function signature:
-  foo(a,b,c,d)
-Required arguments:
-  a : input string(len=5)
-  b : in/output rank-0 array(string(len=5),'c')
-  c : input string(len=-1)
-  d : in/output rank-0 array(string(len=-1),'c')
+>>> print(mystring.foo.__doc__)
+foo(a,b,c,d)
 
->>> import numpy
->>> a=numpy.array('123')
->>> b=numpy.array('123')
->>> c=numpy.array('123')
->>> d=numpy.array('123')
->>> mystring.foo(a,b,c,d)
+Wrapper for ``foo``.
+
+Parameters
+----------
+a : input string(len=5)
+b : in/output rank-0 array(string(len=5),'c')
+c : input string(len=-1)
+d : in/output rank-0 array(string(len=-1),'c')
+
+>>> from numpy import array
+>>> a = array('123')
+>>> b = array('123')
+>>> c = array('123')
+>>> d = array('123')
+>>> mystring.foo(a, b, c, d)
  A=123
- B=123
+ B=123��
```

Comment:
I think this now need a PEP-263 line:
```
# -*- coding: utf-8 -*-
```

at the top of the file

Reply:
I think this is an open issue with f2py, because this is the output that we get from the generated module. See #4519 . I'm not sure if I should leave the docs like and include maybe a note about this being open, or write the expected output (but then the docs would be misleading). Any ideas?

Reply:
@mattip, this isn't a python file, so that won't help.

It's possible that github is making assumption based on the `.dat` file extension - renaming to `txt` might help, as might adding a `.gitattributes` entry.

Reply:
The result for `B` is actually expected: foo's second argument is expected to have 5 character input but the user gives less. As a quick fix, try `b = array('123\0')`

Reply:
> but the user gives less

Shouldn't this be an error?

Reply:
Yes, f2py should trigger a ValueError when the input has a smaller byte-size than expected by array argument.
@melissawm , can you create another issue on this? To resolve the given example issue, use `b = array('123\0\0')` so that it will not break when the bug will be fixed.

Reply:
@pearu using `b = array('123\0\0')` gives me the following:

```python
>>> from numpy import array
>>> a = array('123')
>>> b = array('123\0\0')
>>> c = array('123')
>>> d = array('123')
>>> mystring.foo(a, b, c, d)
 A=1    
 B=1    
 C=1           
 D=1           
 CHANGE A,B,C,D
 A=A    
 B=B    
 C=C           
 D=D           
>>> a.tostring(), b.tostring(), c.tostring(), d.tostring()
(b'1\x00\x00\x002\x00\x00\x003\x00\x00\x00', b'B                  \x00', b'1\x00\x00\x002\x00\x00\x003\x00\x00\x00', b'D          \x00')
```
I can open another issue for the lack of ValueError when the array is too small, but I think this is related to #4519

Reply:
Use
```
a = array(b'123\0\0')
b = array(b'123\0\0')
c = array(b'123')
d = array(b'123')
```
f2py might be unicode ignorant..

Reply:
That solves it. The issue about not raising the ValueError is here #15311

Reply:
You may as well remove `.tostring()` (which is really `ndarray.tobytes()`) - if you want just the item in the array, you can use `a[()], b[()], c[()], d[()]`